### DISCUSSION TITLE
About the challenges of afterEach in interactive test tools

### DISCUSSION THREAD

**Discussion by kasperpeulen** [vitest-dev/vitest#7141] [General]
Hello! Kasper from Storybook here.

At Storybook, we’re exploring adding an `afterEach` hook that runs after each story test execution, similar to how afterEach works in tools like Vitest. However, we’ve encountered some fundamental challenges with incorporating afterEach into an interactive testing tool.

I’m reaching out because I suspect you’ve faced similar issues and thought it would be valuable to align our approaches.

### Using `afterEach` for Cleanup Impedes Interaction

In an interactive testing environment, performing cleanup immediately after test execution can render the test uninteractive. For example, Vitest’s fork of @testing-library/react avoids DOM cleanup in afterEach and instead performs it in beforeEach:
https://github.com/vitest-dev/vitest-browser-react
> Unlike `@testing-library/react`, `vitest-browser-react` performs cleanup of the component before the test begins, allowing you to see the rendered result in your UI. 

But this problem occurs *every time* cleanup is done in `afterEach.` For another example, MSW recommends resetting all handlers in `afterEach`.

```tsx
import { beforeAll, afterEach, afterAll } from 'vitest'
import { server } from './src/mocks/node'
 
beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())
```

In Vitest Browser UI, that would mean that interacting with the test won't hit the mocked endpoints anymore. This can be solved, by doing the reset handlers in a `beforeEach`.

```tsx
beforeEach(() => server.resetHandlers())
```

### Delaying the execution breaks if assertions are used in `afterEach`

At Storybook, we haven’t yet adopted `afterEach` but recently introduced `beforeEach`, which supports a return callback for cleanup:

```tsx
const meta: Meta<typeof Page> = {
  component: Page,
  // 👇 Set the value of Date for every story in the file
  async beforeEach() {
    MockDate.set('2024-02-14');

    // 👇 Reset the Date after each story
    return () => {
      MockDate.reset();
    };
  },
};
```

Instead of running the cleanup callback immediately after the test, we defer it until switching stories. This avoids the interaction issue but creates a new problem: it prevents post-test assertions.

For example, a user trying to assert that console.error wasn’t called in any story wouldn’t see the assertion in Storybook due to the delayed execution:

```tsx
// preview.ts is a place to add global beforeEach in storybook
const preview = {
  async beforeEach() {
    const consoleError = spyOn(console, 'error');    
    
    return () => {
      await expect(consoleError).not.toHaveBeenCalled();      
    }
  }
}
```

While assertions in afterEach are sometimes discouraged, they can be valuable in scenarios like the one above. Other examples include automatically checking for accessibility violations after each test or capturing a DOM/visual snapshot.

### Proposed path forward

So concluding, we think that there is need for 2 distinct stages at the end of testing lifecycle:
1. one for post-test assertions, that happens after the test execution
2. one for cleanup which happens when switching test/stories.

To address this, we are considering the following approach:

1. Introduce `afterEach` to Storybook, but recommend its use for post-test assertions rather than cleanup.
2. Encourage people to do global cleanup in beforeEach, following the pattern used by vitest-browser-react.
3. Maintain the current behavior in storybook of using the return callback from beforeEach for story-specific cleanup, executed when switching stories.

Alternatively, we could introduce a dedicated hook for post-test assertions instead of adopting afterEach. This would avoid the confusion of adding afterEach while advising against its use for cleanup.

We’re curious to hear your thoughts on the best path forward and whether there’s an opportunity to align our approaches.

**Comment by sheremet-va** [comment]
Interesting point. I think it might be unique to projects that have a visual interface where components can be viewed after the fact and cannot be generalized into a standardized built-in hook in a test runner like Vitest unless we can define it in another way ("a hook that should be called to release resources", for example).

`vitest-browser-react` (and other `vitest-browser` packages) also defines a hidden `cleanup` function that we call before removing the iframe, so I can see how this potential hook is different from just `beforeEach` and can be used in different situations.

I think if we want to have a general solution, it's best to recommend `beforeEach`. On the other hand, I am also not against introducing a hook like `onCleanup` or `onReleaseResources` to have an easy to understand interface for any situation that requires release of resources (component cleanup, closing server, clearing timeouts, etc. By default, it's just an alias to `afterEach`), but I am a bit afraid of introducing too much mental complexity with it (or does it actually reduce it? I'm not sure yet) - you need to know _why_ you call a specific function instead of _when_ to call a function - I would love to know your take on that.

**Reply by shilman** [reply]
Good points. I agree that this might be specific to UI tools, but that it's still worth supporting since that's a valid use case.

I don't think `beforeEach` is intuitive. How can I clean up a resource before I've allocated it? How does it work with parallelism where I might want to run 100 `beforeEach`'s concurrently?

I like `onCleanup`. It's explicit and tied to the current test you're running rather than the previous test, which seems problematic. I also like the return callback approach for the same reason. But `onCleanup` is more explicit and doesn't involve any breaking changes.

**Reply by kasperpeulen** [reply]
@sheremet-va I think we should keep executing `beforeEach` and `afterEach` right before and after a test.
Even though that means that the we should change the long standing recommendation to do cleanup in `afterEach`. 
Indeed the name is about `when`, and not `why`. 

This means that on *global* level, you can often just move the `afterEach` to a `beforeEach` to solve the problem described.

But on a test file level, you can not move the `afterEach` to a `beforeEach`:

```ts
import MockDate from 'mockdate';

beforeEach(() => {
  MockDate.set('2024-02-14');
});

afterEach(() => {
  MockDate.reset();
});
```

In this case, you could move the `afterEach` to a global `beforeEach`. Otherwise it might affect other test files (at least in non-isolated mode). And in some sense, it might not even be a bad idea, as it makes sure you never have to think about if you cleaned up a MockDate in a test anymore. 

However, I do think a user should have a way to cleanup state that is created in a specific test or test file, without have to change global code. So in Storybook, we support this:

```ts
import MockDate from 'mockdate';

const MyStory = {
  beforeEach(() => {
    MockDate.set('2024-02-14');

    // in storybook, this callback gets *not* called immediately after the test
    // but only when the switching stories 
    return () => {
      MockDate.reset();    
    }
  });
}
```

So I wonder, if Vitest would consider doing the same:
*  Using `afterEach` as a hook that runs immediately after the story. 
* And the return callback of `beforeEach` that runs after the test, but in browser UI mode, this is delayed to the point that the iframe is destroyed, and not immediately after it.

In this way both tools are consistent, and there is a dedicated cleanup/release resources hook as the return callback of beforeEach.

Although, it is not really self-explaining if you only want to use `beforeEach` for cleanup:

```ts
beforeEach(() => {
  return () => {
    MockDate.reset();
  };
});
```

So it might still be nice to have the `onCleanup` hook for those cases:

```ts
onCleanup(() => {
  MockDate.reset();
});
```

**Reply by sheremet-va** [reply]
> In this way both tools are consistent

The tools will be consistent, but Vitest won't be consistent inside itself - the proposed change introduces differences to how hooks run in the browser and Node.js. We are very strict about keeping the same functionality in any environment.

The returned callback _is_ the `afterEach` callback and we do not plan to change how it works. However, I think we can implement the `onCleanup` hook.

**Reply by sheremet-va** [reply]
Personally, I think for Storybook it might be better to introduce a separate hook like `prepareStory` that has a cleanup callback to avoid mixing use cases

**Reply by kasperpeulen** [reply]
> The tools will be consistent, but Vitest won't be consistent inside itself - the proposed change introduces differences to how hooks run in the browser and Node.js. We are very strict about keeping the same functionality in any environment.

I see what you mean, but from a test runner perspective (on the CLI/CI), it would work the same. Only, when interacting with the test, it would work differently. That said, I do understand that this is (breaking) change that you don't want to make.

> The returned callback is the afterEach callback and we do not plan to change how it works. However, I think we can implement the onCleanup hook.

We are also open to changing the return callback of `beforeEach` to how it works in Vitest (in Storybook 9). 
And implement `onCleanup` instead for this kind of functionality, to stay consistent with Vitest.
